test/unit_tests/braket/circuits/test_circuit.py

```diff
@@ -841,6 +846,546 @@ def test_circuit_to_ir_openqasm(circuit, serialization_properties, expected_ir):
+                        "bit[2] b;",
+                        "qubit[2] q;",
+                        "#pragma braket noise two_qubit_depolarizing(0.1) q[0], q[1]",
+                        "b[0] = measure q[0];",
+                        "b[1] = measure q[1];",
+                    ]
+                ),
+                inputs={},
+            ),
+        ),
+        (
+            Circuit().two_qubit_dephasing(0, 1, probability=0.1),
+            OpenQasmProgram(
+                source="\n".join(
+                    [
+                        "OPENQASM 3.0;",
+                        "bit[2] b;",
+                        "qubit[2] q;",
+                        "#pragma braket noise two_qubit_dephasing(0.1) q[0], q[1]",
+                        "b[0] = measure q[0];",
+                        "b[1] = measure q[1];",
+                    ]
+                ),
+                inputs={},
+            ),
+        ),
+        (
+            Circuit().two_qubit_dephasing(0, 1, probability=0.1),
+            OpenQasmProgram(
+                source="\n".join(
+                    [
+                        "OPENQASM 3.0;",
+                        "bit[2] b;",
+                        "qubit[2] q;",
+                        "#pragma braket noise two_qubit_dephasing(0.1) q[0], q[1]",
+                        "b[0] = measure q[0];",
+                        "b[1] = measure q[1];",
+                    ]
+                ),
+                inputs={},
+            ),
+        ),
+        (
+            Circuit().h(0).sample(observable=Observable.Z(), target=0),
+            OpenQasmProgram(
+                source="\n".join(
+                    [
+                        "OPENQASM 3.0;",
+                        "qubit[1] q;",
+                        "h q[0];",
+                        "#pragma braket result sample z(q[0])",
+                    ]
+                ),
+                inputs={},
+            ),
+        ),
+        (
+            Circuit().h(0).sample(observable=Observable.Z(), target=0),
+            OpenQasmProgram(
+                source="\n".join(
+                    [
+                        "OPENQASM 3.0;",
+                        "qubit[1] q;",
+                        "h q[0];",
+                        "#pragma braket result sample z(q[0])",
+                    ]
+                ),
+                inputs={},
+            ),
+        ),
+        (
+            Circuit().h(0).x(1).density_matrix(target=[0, 1]),
+            OpenQasmProgram(
+                source="\n".join(
+                    [
+                        "OPENQASM 3.0;",
+                        "qubit[2] q;",
+                        "h q[0];",
+                        "x q[1];",
+                        "#pragma braket result density_matrix q[0], q[1]",
+                    ]
+                ),
+                inputs={},
+            ),
+        ),
+    ],
+)
+def test_openqasm_from_ir_circuit(expected_circuit, ir):
```

Comment:
Imo, this test is missing the coolest part of your project. It's pretty simple to make a 1:1 map between BDK circuits and local sim circuits. The exciting part of this code is that it plugs in to the local sim interpreter, allowing complex circuits using greater openqasm scope to be compiled into circuits in the BDK. I'd love to see some tests with subroutines, control flow, custom gate definitions, variables, etc.

Feel free to include that as a separate unit test to explicitly separate between BDK parity and greater functionality

Reply:
I have added few unit tests: https://github.com/aws/amazon-braket-sdk-python/pull/587/files#diff-d1be7e57eeed50bf0ac3974297a346a7c9564c284dedc55b562e11a63a6c1c3aR1480

It was fun to watch it work. : )